homeassistant/components/livebox/__init__.py

```diff
@@ -0,0 +1,166 @@
+        config_entry_id=config_entry.entry_id,
+        identifiers={(DOMAIN, config_entry.data["id"])},
+        manufacturer=config["Manufacturer"],
+        name=config["ProductClass"],
+        model=config["ModelName"],
+        sw_version=config["SoftwareVersion"],
+    )
+
+    for component in COMPONENTS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(config_entry, component)
+        )
+
+    return True
+
+
+async def async_unload_entry(hass, config_entry):
+    """Unload a config entry."""
+
+    for component in COMPONENTS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_unload(config_entry, component)
+        )
+
+    return True
+
+
+class LiveboxData:
+    """Collect datas information from livebox."""
+
+    def __init__(self, config_entry):
+        """Init datas from router."""
+
+        self._box = Sysbus()
+        self._entry = config_entry
+
+    async def async_devices(self, device=None):
+        """Get devices datas."""
+
+        parameters = {"parameters": {"expression": {"wifi": "wifi"}}}
+        if device is not None:
+            parameters = {
+                "parameters": {
+                    "expression": {
+                        "wifi": 'wifi and .PhysAddress=="' + device["PhysAddress"] + '"'
+                    }
+                }
+            }
+        box = await self.async_conn()
+        devices = await box.system.get_devices(parameters)
+        if devices.get("status", {}).get("wifi", {}):
+            return devices["status"]["wifi"]
+        return None
+
+    async def async_infos(self):
+        """Get infos."""
+
+        box = await self.async_conn()
+        infos = await box.system.get_deviceinfo()
+        if infos.get("status", {}):
+            return infos["status"]
+        return None
+
+    async def async_status(self):
+        """Get status."""
+
+        box = await self.async_conn()
+        status = await box.system.get_WANStatus()
+        if status.get("data", {}):
+            return status["data"]
+        return None
+
+    async def async_dsl_status(self):
+        """Get dsl status."""
+
+        box = await self.async_conn()
+        parameters = {"parameters": {"mibs": "dsl", "flag": "", "traverse": "down"}}
+        dsl_status = await box.connection.get_data_MIBS(parameters)
+        if dsl_status.get("status", {}).get("dsl", {}).get("dsl0", {}):
+            return dsl_status["status"]["dsl"]["dsl0"]
+        return None
+
+    async def async_conn(self):
+        """Connect at the livebox router."""
+
+        try:
+            await self._box.open(
+                host=self._entry.data["host"],
+                port=self._entry.data["port"],
+                username=self._entry.data["username"],
+                password=self._entry.data["password"],
+            )
+            if await self._box.get_permissions():  # check connection successful
+                return self._box
```

Comment:
Why return the box? It's an instance attribute so we have access to it in all instance methods.